File: commonplace/instantiate.py
from calendar import monthrange
from datetime import datetime, timedelta
from typing import List, Iterator
from commonplace.models import (Instance, Moment, MomentDateTime, Recurrence, RecurrenceType, RecurringMoment, SingleMoment,
                                WorkState)
from commonplace.util import epoch_week, with_end_of_day, with_weekday


def generate_instances(moments: List[Moment], start: datetime, end: datetime, incl_subs=True, predicate=None):
    return [i for m in moments for i in generate_instances_of_moment(m, start, end, incl_subs=incl_subs, predicate=predicate)]


def generate_instances_of_moment(moment: Moment, start: datetime, end: datetime, incl_subs=True, predicate=None):
    end = with_end_of_day(end)
    instances: List[Instance] = [i for i in create_instances(moment, start, end) if predicate is None or predicate(i)]
    if incl_subs:
        for inst in instances:
            inst.sub_instances = generate_instances(moment.sub_moments, inst.start, inst.end, incl_subs, predicate)

    return instances


def create_instances(moment: Moment, start: datetime, end: datetime):
    if isinstance(moment, SingleMoment):
        return create_single_moment_instances(moment, start, end)

    if isinstance(moment, RecurringMoment):
        return create_recurring_moment_instances(moment, start, end)

    return []


def create_single_moment_instances(moment: SingleMoment, start: datetime, end: datetime):
    inst_start, inst_end = intersect(start, end, moment.start, moment.end)
    if not inst_start:
        return []

    inst = create_instance(moment, inst_start, inst_end, ends_in_range=moment.end is not None and moment.end.dt <= inst_end)
    return [inst]


def create_recurring_moment_instances(moment: RecurringMoment, start: datetime, end: datetime):
    return [
        create_instance(moment, inst_start, with_end_of_day(inst_start), ends_in_range=True)
        for inst_start in generate_recurring(moment.recurrence, start, end)
    ]


def create_instance(moment: Moment, inst_start, inst_end, ends_in_range):
    return Instance(
        name=moment.name,
        start=inst_start,
        end=inst_end,
        time_of_day=moment.time_of_day.dt.time() if moment.time_of_day else None,
        priority=moment.priority,
        category=moment.category,
        done=moment.work_state == WorkState.DONE,
        work_state=moment.work_state,
        ends_in_range=ends_in_range,
        origin_doc_pos=moment.doc_pos,
    )


def intersect(start1: datetime, end1: datetime, start2: MomentDateTime | None, end2: MomentDateTime | None):
    """ returns the start and end of the intersection of two datetime ranges or (None,None) if they don't overlap. """
    upper_start = upper_bound(start1, start2.dt if start2 else None)
    lower_end = lower_bound(end1, end2.dt if end2 else None)
    if lower_end < upper_start:
        # No overlap
        return None, None

    return upper_start, lower_end


def lower_bound(t1: datetime, t2: datetime) -> datetime | None:  # pylint: disable=invalid-name
    """ returns the earlier of two datetimes. A nil time is considered infinitely late so won't be
    used if the other time is not nil. """
    if t1 is None:
        # if t2 is also None, we return None
        return t2

    if t2 is None:
        return t1

    return t1 if t1 < t2 else t2


def upper_bound(t1: datetime, t2: datetime) -> datetime | None:  # pylint: disable=invalid-name
    """ returns the later of two datetimes. A nil time is considered infinitely early so won't be
    used if the other time is not nil. """
    if t1 is None:
        # if t2 is also None, we return None
        return t2

    if t2 is None:
        return t1

    return t1 if t1 > t2 else t2


def generate_recurring(recurrence: Recurrence, start: datetime, end: datetime) -> Iterator[datetime]:
    cur = start - timedelta(days=1)
    while True:
        cur = next_recurring(recurrence, cur)
        if cur <= end:
            yield cur
        else:
            break


def next_recurring(recurrence: Recurrence, after: datetime):
    nexts = {
        RecurrenceType.DAILY: next_daily,
        RecurrenceType.WEEKLY: next_weekly,
        RecurrenceType.BI_WEEKLY: lambda a, rf: next_nweekly(2, a, rf),
        RecurrenceType.TRI_WEEKLY: lambda a, rf: next_nweekly(3, a, rf),
        RecurrenceType.QUADRI_WEEKLY: lambda a, rf: next_nweekly(4, a, rf),
        RecurrenceType.MONTHLY: next_monthly,
        RecurrenceType.YEARLY: next_yearly,
    }
    return nexts[recurrence.recurrence_type](after, recurrence.ref_date.dt)


def next_daily(after: datetime, _: datetime):
    return after + timedelta(days=1)


def next_weekly(after: datetime, ref_date: datetime):
    date = with_weekday(after, ref_date.weekday())
    if date <= after:
        date += timedelta(days=7)
    return date


def next_nweekly(nth: int, after: datetime, ref_date: datetime):
    date = with_weekday(after, ref_date.weekday())
    if date <= after:
        date += timedelta(days=7)

    offset = (epoch_week(date) - epoch_week(ref_date)) % nth
    if offset > 0:
        date += timedelta(days=7 * (nth - offset))
    elif offset < 0:
        date += timedelta(days=-7 * offset)

    return date


def next_monthly(after: datetime, ref_date: datetime):
    date = datetime(
        year=after.year,
        month=after.month,
        day=min(ref_date.day, last_day_of_month(after.year, after.month)),
        hour=0,
        minute=0,
        second=0,
        microsecond=0,
    ).astimezone(tz=None)

    if date <= after:
        year = date.year + 1 if date.month == 12 else date.year
        month = date.month % 12 + 1
        date = date.replace(
            year=year,
            month=month,
            day=min(ref_date.day, last_day_of_month(year, month)),
        )

    return date


def last_day_of_month(year, month):
    return monthrange(year, month)[1]


def next_yearly(after: datetime, ref_date: datetime):
    date = datetime(
        year=after.year,
        month=ref_date.month,
        day=ref_date.day,
        hour=0,
        minute=0,
        second=0,
        microsecond=0,
    ).astimezone(tz=None)

    if date <= after:
        date = date.replace(year=date.year + 1)
    return date

File: core/lib/src/instantiate.ts
import { addDays, addYears, endOfDay, getDay, getDaysInMonth, isAfter, isBefore, setDay } from 'date-fns'
import { Instance, isRecurringMoment, isSingleMoment, Moment, MomentDateTime, Recurrence, RecurrenceType, RecurringMoment, SingleMoment, WorkState } from './models'
import { epochWeek } from './util'

interface GenerateOptions {
    inclSubs?: boolean
    predicate?: (i: Instance) => boolean
}

export function generateInstances(moments: Moment[], start: Date, end: Date, options: GenerateOptions = { inclSubs: true }) {
    return moments.flatMap(m => generateInstancesOfMoment(m, start, end, options))
}

export function generateInstancesOfMoment(moment: Moment, start: Date, end: Date, options: GenerateOptions = { inclSubs: true }) {
    end = endOfDay(end)
    const instances: Instance[] = createInstances(moment, start, end).filter(i => !options.predicate || options.predicate(i))
    if (options.inclSubs === undefined || options.inclSubs) {
        instances.forEach(inst => {
            inst.subInstances = generateInstances(moment.subMoments, inst.start, inst.end, options)
        })
    }

    return instances
}

function createInstances(moment: Moment, start: Date, end: Date) {
    if (isSingleMoment(moment)) {
        return createSingleMomentInstances(moment as SingleMoment, start, end)
    }

    if (isRecurringMoment(moment)) {
        return createRecurringMomentInstances(moment as RecurringMoment, start, end)
    }

    return []
}

function createSingleMomentInstances(moment: SingleMoment, start: Date, end: Date) {
    const [instStart, instEnd] = intersect(start, end, moment.start, moment.end)
    if (!instStart || !instEnd) {
        return []
    }

    const endsInRange = moment.end !== null && !isAfter(moment.end.dt, instEnd)
    const inst = createInstance(moment, instStart, instEnd, endsInRange)
    return [inst]
}

function createRecurringMomentInstances(moment: RecurringMoment, start: Date, end: Date) {
    const instances = []
    for (const instStart of generateRecurring(moment.recurrence, start, end)) {
        instances.push(createInstance(moment, instStart, endOfDay(instStart), true))
    }
    return instances
}

function createInstance(moment: Moment, instStart: Date, instEnd: Date, endsInRange: boolean): Instance {
    return {
        name: moment.name,
        start: instStart,
        end: instEnd,
        timeOfDay: moment.timeOfDay?.dt ?? null,
        priority: moment.priority,
        category: moment.category,
        done: moment.workState === WorkState.DONE,
        workState: moment.workState,
        endsInRange,
        originDocPos: moment.docPos,
        subInstances: []
    }
}

/**
 * returns the start and end of the intersection of two Date ranges or (None,None) if they don't overlap.
 */
function intersect(start1: Date, end1: Date, start2: MomentDateTime | null, end2: MomentDateTime | null): [Date | null, Date | null] {
    const latestStart = latest(start1, start2?.dt ?? null)
    const earliestEnd = earliest(end1, end2?.dt ?? null)
    if (latestStart && earliestEnd && isBefore(earliestEnd, latestStart)) {
        // No overlap
        return [null, null]
    }

    return [latestStart, earliestEnd]
}

/**
 *  returns the earlier of two Dates. A nil time is considered infinitely late so won't be used if the other time is not nil.
 */
function earliest(t1: Date | null, t2: Date | null): Date | null {
    if (t1 === null) {
        // if t2 is also null, we return null
        return t2
    }

    if (t2 === null) {
        return t1
    }

    return isBefore(t1, t2) ? t1 : t2
}

/**
 * returns the later of two Dates. A nil time is considered infinitely early so won't be used if the other time is not nil.
 */
function latest(t1: Date | null, t2: Date | null): Date | null {
    if (t1 === null) {
        // if t2 is also None, we return None
        return t2
    }

    if (t2 === null) {
        return t1
    }
    return isAfter(t1, t2) ? t1 : t2
}

function* generateRecurring(recurrence: Recurrence, start: Date, end: Date): Generator<Date> {
    let cur = addDays(start, -1)
    while (true) {
        cur = nextRecurring(recurrence, cur)
        if (!isAfter(cur, end)) {
            yield cur
        }
        else {
            break
        }
    }
}

function nextRecurring(recurrence: Recurrence, after: Date) {
    switch (recurrence.recurrenceType) {
        case RecurrenceType.DAILY: return nextDaily(after, recurrence.refDate.dt)
        case RecurrenceType.WEEKLY: return nextWeekly(after, recurrence.refDate.dt)
        case RecurrenceType.BIWEEKLY: return nextNweekly(2, after, recurrence.refDate.dt)
        case RecurrenceType.TRIWEEKLY: return nextNweekly(3, after, recurrence.refDate.dt)
        case RecurrenceType.QUADRIWEEKLY: return nextNweekly(4, after, recurrence.refDate.dt)
        case RecurrenceType.MONTHLY: return nextMonthly(after, recurrence.refDate.dt)
        case RecurrenceType.YEARLY: return nextYearly(after, recurrence.refDate.dt)
        default: throw Error(`Unknown recurrence type ${recurrence.recurrenceType}`)
    }
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function nextDaily(after: Date, _: Date) {
    return addDays(after, 1)
}

function nextWeekly(after: Date, refDate: Date) {
    let date = setDay(after, getDay(refDate))
    if (!isAfter(date, after)) {
        date = addDays(date, 7)
    }
    return date
}

function nextNweekly(nth: number, after: Date, refDate: Date) {
    let date = setDay(after, getDay(refDate))
    if (!isAfter(date, after)) {
        date = addDays(date, 7)
    }

    const offset = (epochWeek(date) - epochWeek(refDate)) % nth
    if (offset > 0) {
        date = addDays(date, 7 * (nth - offset))
    }
    else if (offset < 0) {
        date = addDays(date, -7 * offset)
    }

    return date
}

function nextMonthly(after: Date, refDate: Date) {
    let date = new Date(
        after.getFullYear(),
        after.getMonth(),
        Math.min(refDate.getDate(), getDaysInMonth(after)),
        0,
        0,
        0,
        0
    )

    if (!isAfter(date, after)) {
        const year = date.getMonth() === 11 ? date.getFullYear() + 1 : date.getFullYear()
        const month = (date.getMonth() + 1) % 12
        date = new Date(
            year,
            month,
            Math.min(refDate.getDate(), getDaysInMonth(new Date(year, month, 1))),
            0,
            0,
            0,
            0
        )
    }

    return date
}

function nextYearly(after: Date, refDate: Date) {
    let date = new Date(
        after.getFullYear(),
        refDate.getMonth(),
        refDate.getDate(),
        0,
        0,
        0,
        0
    )

    if (!isAfter(date, after)) {
        date = addYears(date, 1)
    }
    return date
}
